File: server/shopifyOAuth.ts
import crypto from "crypto";

export const DEFAULT_SHOPIFY_SCOPES = "read_products,write_products,read_inventory,write_inventory,read_locations";
export const DEFAULT_APP_BASE_URL = "https://snapsyncai.co.uk";

const SHOPIFY_SHOP_DOMAIN_RE = /^[a-zA-Z0-9][a-zA-Z0-9-]*\.myshopify\.com$/;
const STATE_TTL_MS = 10 * 60 * 1000;

type ShopifyOAuthState = {
  userId: string;
  nonce: string;
  ts: number;
};

export type ShopifyOAuthStateResult =
  | { ok: true; userId: string }
  | { ok: false; reason: "malformed" | "invalid_signature" | "expired" };

function hmacSha256Hex(secret: string, message: string) {
  return crypto.createHmac("sha256", secret).update(message).digest("hex");
}

function timingSafeEqualHex(a: string, b: string) {
  if (!/^[a-f0-9]+$/i.test(a) || !/^[a-f0-9]+$/i.test(b) || a.length !== b.length) {
    return false;
  }

  return crypto.timingSafeEqual(Buffer.from(a, "hex"), Buffer.from(b, "hex"));
}

export function normalizeShopifyDomain(rawShop: string) {
  const domain = rawShop
    .trim()
    .toLowerCase()
    .replace(/^https?:\/\//, "")
    .replace(/\/.*$/, "");
  return domain.includes(".myshopify.com") ? domain : `${domain}.myshopify.com`;
}

export function isValidShopifyDomain(shop: string) {
  return SHOPIFY_SHOP_DOMAIN_RE.test(shop);
}

export function createShopifyOAuthState(userId: string, secret: string, now = Date.now()) {
  const payload: ShopifyOAuthState = {
    userId,
    nonce: crypto.randomBytes(16).toString("hex"),
    ts: now,
  };
  const encoded = Buffer.from(JSON.stringify(payload), "utf8").toString("base64url");
  return `${encoded}.${hmacSha256Hex(secret, encoded)}`;
}

export function verifyShopifyOAuthState(
  state: string,
  secret: string,
  now = Date.now(),
): ShopifyOAuthStateResult {
  const [encoded, signature] = state.split(".");
  if (!encoded || !signature) return { ok: false, reason: "malformed" };

  const expected = hmacSha256Hex(secret, encoded);
  if (!timingSafeEqualHex(signature, expected)) {
    return { ok: false, reason: "invalid_signature" };
  }

  try {
    const payload = JSON.parse(Buffer.from(encoded, "base64url").toString("utf8")) as ShopifyOAuthState;
    if (!payload.userId || !payload.nonce || typeof payload.ts !== "number") {
      return { ok: false, reason: "malformed" };
    }
    if (now - payload.ts > STATE_TTL_MS) {
      return { ok: false, reason: "expired" };
    }
    return { ok: true, userId: payload.userId };
  } catch {
    return { ok: false, reason: "malformed" };
  }
}

export function verifyShopifyHmac(
  query: Record<string, unknown>,
  secret: string,
) {
  const hmac = query.hmac;
  const provided = Array.isArray(hmac) ? hmac[0] : hmac;
  if (typeof provided !== "string") return false;

  const entries: [string, string][] = [];
  for (const [key, value] of Object.entries(query)) {
    if (key === "hmac" || key === "signature") continue;
    const values = Array.isArray(value) ? value : [value];
    for (const item of values) {
      if (item === undefined) continue;
      entries.push([key, String(item)]);
    }
  }

  const message = entries
    .sort(([aKey, aValue], [bKey, bValue]) => aKey.localeCompare(bKey) || aValue.localeCompare(bValue))
    .map(([key, value]) => `${key}=${value}`)
    .join("&");

  const expected = hmacSha256Hex(secret, message);
  return timingSafeEqualHex(provided, expected);
}

export function buildShopifyOAuthAuthorizeUrl({
  shop,
  apiKey,
  scopes,
  redirectUri,
  state,
}: {
  shop: string;
  apiKey: string;
  scopes: string;
  redirectUri: string;
  state: string;
}) {
  const params = new URLSearchParams({
    client_id: apiKey,
    scope: scopes,
    redirect_uri: redirectUri,
    state,
  });

  return `https://${shop}/admin/oauth/authorize?${params.toString()}`;
}

export function getShopifyOAuthConfig() {
  return {
    apiKey: process.env.SHOPIFY_API_KEY || process.env.SHOPIFY_CLIENT_ID || "",
    apiSecret: process.env.SHOPIFY_API_SECRET || process.env.SHOPIFY_CLIENT_SECRET || "",
    scopes: process.env.SHOPIFY_SCOPES || DEFAULT_SHOPIFY_SCOPES,
    appBaseUrl: (process.env.APP_BASE_URL || DEFAULT_APP_BASE_URL).replace(/\/$/, ""),
  };
}

File: api-py/app/services/shopify_oauth.py
"""Shopify OAuth helpers — port of `server/shopifyOAuth.ts`.

State tokens and callback HMACs must verify values produced by Express (and
the other way around) so a rewrite that splits start/callback across backends
does not drop the user in an `invalid_state_*` loop.
"""

from __future__ import annotations

import hashlib
import hmac
import json
import re
import secrets
import time
from typing import Literal
from urllib.parse import urlencode

from app.config import Settings

DEFAULT_SHOPIFY_SCOPES = (
    "read_products,write_products,read_inventory,write_inventory,read_locations"
)
DEFAULT_APP_BASE_URL = "https://snapsyncai.co.uk"

SHOPIFY_SHOP_DOMAIN_RE = re.compile(r"^[a-zA-Z0-9][a-zA-Z0-9-]*\.myshopify\.com$")
STATE_TTL_MS = 10 * 60 * 1000

ShopifyOAuthStateReason = Literal["malformed", "invalid_signature", "expired"]


def hmac_sha256_hex(secret: str, message: str) -> str:
    return hmac.new(secret.encode("utf-8"), message.encode("utf-8"), hashlib.sha256).hexdigest()


def timing_safe_equal_hex(left: str, right: str) -> bool:
    if not re.fullmatch(r"[a-f0-9]+", left, re.I) or not re.fullmatch(r"[a-f0-9]+", right, re.I):
        return False
    if len(left) != len(right):
        return False
    return hmac.compare_digest(bytes.fromhex(left), bytes.fromhex(right))


def normalize_shopify_domain(raw_shop: str) -> str:
    domain = re.sub(r"^https?://", "", raw_shop.strip().lower())
    domain = re.sub(r"/.*$", "", domain)
    return domain if ".myshopify.com" in domain else f"{domain}.myshopify.com"


def is_valid_shopify_domain(shop: str) -> bool:
    return bool(SHOPIFY_SHOP_DOMAIN_RE.match(shop))


def _b64url_encode(data: bytes) -> str:
    from base64 import urlsafe_b64encode

    return urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def _b64url_decode(text: str) -> bytes:
    from base64 import urlsafe_b64decode

    pad = "=" * (-len(text) % 4)
    return urlsafe_b64decode(text + pad)


def create_shopify_oauth_state(user_id: str, secret: str, now: int | None = None) -> str:
    payload = {
        "userId": user_id,
        "nonce": secrets.token_hex(16),
        "ts": int(time.time() * 1000) if now is None else now,
    }
    encoded = _b64url_encode(json.dumps(payload, separators=(",", ":")).encode("utf-8"))
    return f"{encoded}.{hmac_sha256_hex(secret, encoded)}"


def verify_shopify_oauth_state(
    state: str,
    secret: str,
    now: int | None = None,
) -> dict:
    encoded, _, signature = state.partition(".")
    if not encoded or not signature:
        return {"ok": False, "reason": "malformed"}

    expected = hmac_sha256_hex(secret, encoded)
    if not timing_safe_equal_hex(signature, expected):
        return {"ok": False, "reason": "invalid_signature"}

    try:
        payload = json.loads(_b64url_decode(encoded).decode("utf-8"))
        ts = payload.get("ts")
        # `bool` is an `int` subclass; Express checks `typeof ts === "number"`.
        if (
            not payload.get("userId")
            or not payload.get("nonce")
            or type(ts) not in (int, float)
        ):
            return {"ok": False, "reason": "malformed"}
        clock = int(time.time() * 1000) if now is None else now
        if clock - payload["ts"] > STATE_TTL_MS:
            return {"ok": False, "reason": "expired"}
        return {"ok": True, "userId": payload["userId"]}
    except (ValueError, json.JSONDecodeError, UnicodeDecodeError):
        return {"ok": False, "reason": "malformed"}


def verify_shopify_hmac(query: dict, secret: str) -> bool:
    hmac_value = query.get("hmac")
    provided = hmac_value[0] if isinstance(hmac_value, list) else hmac_value
    if not isinstance(provided, str):
        return False

    entries: list[tuple[str, str]] = []
    for key, value in query.items():
        if key in ("hmac", "signature"):
            continue
        values = value if isinstance(value, list) else [value]
        for item in values:
            if item is None:
                continue
            entries.append((key, str(item)))

    entries.sort(key=lambda pair: (pair[0], pair[1]))
    message = "&".join(f"{key}={value}" for key, value in entries)
    return timing_safe_equal_hex(provided, hmac_sha256_hex(secret, message))


def build_shopify_oauth_authorize_url(
    *,
    shop: str,
    api_key: str,
    scopes: str,
    redirect_uri: str,
    state: str,
) -> str:
    params = urlencode(
        {
            "client_id": api_key,
            "scope": scopes,
            "redirect_uri": redirect_uri,
            "state": state,
        }
    )
    return f"https://{shop}/admin/oauth/authorize?{params}"


def shopify_oauth_config(settings: Settings) -> dict[str, str]:
    return {
        "api_key": settings.shopify_client_id_resolved or "",
        "api_secret": settings.shopify_client_secret_resolved or "",
        "scopes": settings.shopify_scopes or DEFAULT_SHOPIFY_SCOPES,
        "app_base_url": (settings.app_base_url or DEFAULT_APP_BASE_URL).rstrip("/"),
    }
